### PR TITLE
introduce incremental builds

### PR DESCRIPTION
**Pull request by sosiouxme**
This introduces incremental builds, meaning only rebuilding images where the upstream repo has changed from what's in dist-git.

### CODE REVIEW

**Comment by sosiouxme** [comment]
@adammhaile @tbielawa now that doozer release auto-increment works, this is running great over at https://buildvm.openshift.eng.bos.redhat.com:8443/job/Experimental/job/lmeyer/job/build%252Fscan-sources/  (last time it failed only because it was trying to build 4.0 signed... which i fixed)
So, ready for review.

File: jobs/build/incremental/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                to: "${MAIL_LIST_FAILURE}",
+                from: "aos-cicd@redhat.com",
+                replyTo: 'aos-team-art@redhat.com',
+                subject: subject,
+                body: """\
+Encoutered an error ${stage}while running ${env.JOB_NAME}:
+${err}
+
+Jenkins job: ${env.BUILD_URL}
+Console output: ${env.BUILD_URL}console
```
Comment (tbielawa) [inline jobs/build/incremental/Jenkinsfile]: Why extra 'console' at end?

File: jobs/build/incremental/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+                    [
+                        name: 'MOCK',
+                        description: 'Pick up changed job parameters and then exit',
+                        $class: 'BooleanParameterDefinition',
+                        defaultValue: false
+                    ],
+                ]
+            ]
+        ]
+    )
```
Comment (tbielawa) [inline jobs/build/incremental/Jenkinsfile]: I like that you fixed all of the indentation

File: jobs/build/incremental/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+    // User can have the job end if this is just a run to pick up parameter changes
+    // (which Jenkins discovers by running the job).
+    if (MOCK.toBoolean()) {
+        currentBuild.description = "Ran in mock mode"
```
Comment (tbielawa) [inline jobs/build/incremental/Jenkinsfile]: :birthday:

File: jobs/build/incremental/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+                short_images.add("...")
+            }
+            currentBuild.description = "Changes to: " + short_images.join(", ")
+
+            // also determine child images
+            yamlData = readYaml text: buildlib.doozer(
+                "${doozer_opts} images:show-tree --yml", [capture: true]
+            )
+            Closure gather_children
+            gather_children = { all, yaml, changed, gather ->
```
Comment (tbielawa) [inline jobs/build/incremental/Jenkinsfile]: I wouldn't say I'm (or any ARTist) especially 1337 with groovy right now. Please add some flavor text in & around this.

File: jobs/build/incremental/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+            def failed_msg = ""
+
+            echo "Error building images: ${err}"
+            currentBuild.description = "${err} during build step"
+
+            failed_builds = buildlib.get_failed_builds(DOOZER_WORKING)
+            if (failed_builds) {
+                failed_msg = "The following build(s) failed:\n"
+                failed_builds.each { img, reason -> failed_msg += "${img}: ${reason} \n" }
+                println failed_msg
```
Comment (tbielawa) [inline jobs/build/incremental/Jenkinsfile]: Sometimes you do `echo`, sometimes `println`, why is that?

File: jobs/build/incremental/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+    def stage = params.stage ? "in stage '${params.stage}' " : ""
+    def extra_body = params.get("extra_body", "")
+    try {
+        mail(
+                to: "${MAIL_LIST_FAILURE}",
+                from: "aos-cicd@redhat.com",
+                replyTo: 'aos-team-art@redhat.com',
+                subject: subject,
+                body: """\
+Encoutered an error ${stage}while running ${env.JOB_NAME}:
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply jobs/build/incremental/Jenkinsfile]: Nope,` ${stage}` supplies the space if it's defined at all.

File: jobs/build/incremental/Jenkinsfile (hunk 1 above)
Comment (sosiouxme) [reply jobs/build/incremental/Jenkinsfile]: It is a link to the console output of the job. The job URL ends in `/`

File: jobs/build/incremental/Jenkinsfile (hunk 4 above)
Comment (sosiouxme) [reply jobs/build/incremental/Jenkinsfile]: Can do... is it not evident that it's a recursive closure for traversing the child tree?

File: jobs/build/incremental/Jenkinsfile (hunk 5 above)
Comment (sosiouxme) [reply jobs/build/incremental/Jenkinsfile]: I'm actually a clueless groovy n00b. I don't know if there's even a difference.

File: jobs/build/incremental/Jenkinsfile (hunk 4 above)
Comment (tbielawa) [reply jobs/build/incremental/Jenkinsfile]: Uh. Kinda.

File: jobs/build/incremental/Jenkinsfile (hunk 5 above)
Comment (sosiouxme) [reply jobs/build/incremental/Jenkinsfile]: https://stackoverflow.com/questions/42149652/println-in-call-method-of-vars-foo-groovy-works-but-not-in-method-in-class seems relevant if i could understand it a little better.

I don't think it matters but I'm making them all `echo` for consistency.

File: jobs/build/incremental/Jenkinsfile (hunk 4 above)
Comment (sosiouxme) [reply jobs/build/incremental/Jenkinsfile]: Better?

Comment (tbielawa) [reply jobs/build/incremental/Jenkinsfile]: fantastic

**Comment by sosiouxme** [comment]
@adammhaile you want to hold this open to look at it, or just wing it in there and try it out? It's net new so not much risk...

**Comment by adammhaile** [comment]
@sosiouxme PUSH IT :) Looks good

**Pull request merged**